src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,386 @@
+                            batch_records += current_records;
+                        }
+
+                        if data_slice_batch_info.new_file_path.is_none() {
+                            data_slice_batch_info.new_file_path = Some(compact_dir_path.join(
+                                Self::get_random_operation_name_with_prefix(
+                                    format!("{dataset_name}-").as_str(),
+                                ),
+                            ));
+                        }
+                        data_slice_batch_info.prev_offset = add_data_event.prev_offset;
+                        data_slice_batch_info.new_offset_interval = Some(OffsetInterval {
+                            start: output_slice.offset_interval.start,
+                            end: new_end_offset_interval,
+                        });
+                    }
+                }
+                _ => continue,
+            }
+        }
+        if !data_slice_batch_info.data_slices_batch.is_empty() {
+            data_slice_batches.push(data_slice_batch_info);
+        }
+
+        Ok(ChainFilesInfo {
+            data_slice_batches,
+            block_file_urls,
+            new_head: new_head.unwrap(),
+        })
+    }
+
+    async fn write_file(
+        &self,
+        data_slice_batches: &Vec<DataSliceBatchInfo>,
+    ) -> Result<(), CompactError> {
+        let ctx = SessionContext::new();
+
+        for data_slice_batch_info in data_slice_batches {
+            let data_frame = ctx
+                .read_parquet(
+                    data_slice_batch_info.data_slices_batch.clone(),
+                    datafusion::execution::options::ParquetReadOptions {
+                        file_extension: "",
+                        ..Default::default()
+                    },
+                )
+                .await
+                .int_err()?;
+
+            data_frame
+                .write_parquet(
+                    data_slice_batch_info
+                        .new_file_path
+                        .as_ref()
+                        .unwrap()
+                        .to_str()
+                        .unwrap(),
+                    datafusion::dataframe::DataFrameWriteOptions::new()
+                        .with_single_file_output(true),
+                    None,
+                )
+                .await
+                .int_err()?;
+        }
+
+        Ok(())
+    }
+
+    fn create_run_compact_dir(&self) -> Result<PathBuf, CompactError> {
+        let compact_dir_path = self.run_info_dir.join("compact");
+        fs::create_dir_all(&compact_dir_path).int_err()?;
+        Ok(compact_dir_path)
+    }
+
+    async fn replace_data_files(
+        &self,
+        dataset: Arc<dyn Dataset>,
+        chain_files_info: &ChainFilesInfo,
+    ) -> Result<(Vec<Url>, Multihash), CompactError> {
+        let chain = dataset.as_metadata_chain();
+
+        let mut new_head = chain_files_info.new_head.clone();
```

Comment:
So that was a previous block hash before first `AddData`. Not a "new_head" from gathering point of view.